### ISSUE TITLE
Redis Queue: How to prevent chained Jobs from running asynchronously

### ISSUE THREAD

**Issue by ShubhamRathi** [rq/rq#1097]
I'm trying to get 3 Jobs to run after each other in the sequence:

`Job1 -> Job2 -> Job3`
These 3 jobs are defined in operations.py:

```
def Job1(x):
    return x

def Job2(x):
    return x * x

def Job3(x):
    print(x)
```
I'm calling these jobs in script.py with the rq worker running:

```
from redis import Redis
from rq import Queue
from operation import Job1, Job2, Job3

redis_conn = Redis()
q = Queue(connection=redis_conn)

for num in [1,2,3,4,5,6,7,8]:
    j1 = q.enqueue(Job1, num)
    j2 = q.enqueue(Job2, j1.result, depends_on = j1)
    j3 = q.enqueue(Job3, depends_on = j2)
```
As per the [documentation](https://python-rq.org/docs/), I expect j3 to wait for j2 which in turn should wait for j1 to finish execution. But, this probably isnt happening. The jobs seem to be running async. I say so because the redis worker gives this as an error:

```
File "./operation.py", line 5, in Job2
    return x * x
TypeError: unsupported operand type(s) for *: 'NoneType' and 'NoneType'
```
j2 instead of waiting for j1's result is async also firing up and since by then j1's result isnt ready, j1.result is None which is passed to j2. Whats wrong with my approach? Why arent the jobs running sequentially?

**Comment by selwin** [comment]
The following is not something you can do:
```python
j1 = q.enqueue(Job1, num)
j2 = q.enqueue(Job2, j1.result, depends_on = j1)
```
When `j2` is enqueued, `j1.result` has not been computed, which is why `Job2` receives `None` as its argument.

What you can do is modify your `J1` such that it enqueues `Job2` with the result of it's computation.

**Comment by ShubhamRathi** [comment]
I see. Its quite misleading to then state that the jobs can be chained. `depends_on ` attribute is more or less useless if the user has to arrange the jobs. Or maybe I have mis understood? Could you please share an example which involves the use of the `depends_on` parameter.

**Comment by selwin** [comment]
> Its quite misleading to then state that the jobs can be chained

Why is it misleading? Jobs can be chained, no more and no less.

>  Could you please share an example which involves the use of the depends_on parameter.

Here's one to generate a report and after it's done, sends the report via email:
```python
job_1 = queue.enqueue(generate_yearly_report)
job_2 = queue.enqueue(deliver_report_via_email, depends_on=job_1`
```

**Comment by ShubhamRathi** [comment]
Yes, I did see that example, but since its not a working implementation - I'm taking it on face value that it works. As I have demonstrated in my example, the jobs are not sequential and in fact run parallelly. If the jobs were chained (running consecutively), the second job should not start running until the first job has finished. Hence the doubt. 
To simulate chained behavior, this is what I'll instead have to do:

```
for num in [1,2,3,4,5,6,7,8]:
    j1 = q.enqueue(Job1, num)
    while j1.result:
         j2 = q.enqueue(Job2, j1.result)
         while j2.result:
              j3 = q.enqueue(Job3)
```

As seen, there is no use of the `depends_on` parameter here and the jobs would still be chained. My expectation from the `depends_on` parameter is that without the while loop, the jobs are simulated to run sequentially.

**Comment by selwin** [comment]
They do run sequentially. Please check your worker logs and paste the result here.

Your code wouldn't run because `j1.result` is `None` when it's enqueued.
![rq 2019-06-08 20-40-48](https://user-images.githubusercontent.com/188658/59148071-f5973680-8a2d-11e9-988a-db36e2aa044d.jpg)

**Comment by ShubhamRathi** [comment]
I agree that code wouldn't run because ` j1.result` is None during enqueue. But the job (j2) isnt to run when `j1.result` is `None` because I have enqueued it to run AFTER j1 finishes execution (`depends_on = j1`). 
I guess there is a difference of understanding here. My understanding is that j2 is enqueued & executed after j1. But to what I understand from your explanations, j2 is enqueued at the same time as j1, but just that its executed later (after j1). Which is also plausible. 
But the behavior that I want to simulate is that result from j1 decides the nature of execution of j2. To do so, it is necessary that j2 is _enqueued and executed_ after j1. How do I achieve this? One way ofcourse is the one I've showed earlier, using while loops to wait till the result becomes available. This is a tad bit inefficient as the program is polling on `j1.result` continuously. Is there a better way?
